fdbserver/TimedKVCache.h

```diff
@@ -0,0 +1,121 @@
+/**
+ * TimedKVCache.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FDBSERVER_TIMED_KVCACHE_H
+#define FDBSERVER_TIMED_KVCACHE_H
+
+#pragma once
+
+#include <list>
+#include <memory>
+#include <unordered_map>
+
+#include "flow/flow.h"
+#include "flow/Error.h"
+
+/**
+ * @class TimedKVCache
+ * @brief A key/value based cache that expires its key after a given time
+ */
+template <typename K, typename V>
+class TimedKVCache {
+public:
+	using key_t = K;
+	using value_t = V;
+
+	const double CACHE_EXPIRING_TIME_LENGTH;
+
+	explicit TimedKVCache(const double& expiringTimeLength) : CACHE_EXPIRING_TIME_LENGTH(expiringTimeLength) {}
+
+	/**
+	 * Add a new key/value pair to the cache
+	 * @param key
+	 * @param value
+	 */
+	void add(const key_t& key, const value_t& value) {
+		double now = g_network->now();
```

Comment:
Was there any discussion on driving this based on real clock time, versus version time?

Reply:
For simulation, we want to use `now` for determinism. Using real time clock for production probably is not very different from `now` -- the cache entry may expire a little late, which is not a big deal.

Reply:
For what it's worth, I believe `timer` is also deterministic in simulation.

Reply:
`timer` is deterministic yet it adds a determined random value in simulation.

Reply:
My question was not about using now() vs timer(), but instead feeding the highest observed version from any request into the TimedKVCache structure, and making TimedKVCache keep entries around for 1.5million versions instead.  Although it'd be more complicated to thread through this notion of time, it also means that you'd be protected from any clock drift/jumps/etc. on resolvers and tlogs.  I'm not strongly opposed to now(), but I think it is worth considering that this is, I think, the first dependency on clock time that we have in transaction processing (outside of the master).